File: backend/src/llm/model-manager.ts
// backend/src/llm/model-manager.ts
import { BedrockRuntimeClient, InvokeModelCommand } from '@aws-sdk/client-bedrock-runtime';

export interface ModelConfig {
  provider: 'openai' | 'anthropic' | 'bedrock';
  modelId: string;
  maxTokens?: number;
  temperature?: number;
  topP?: number;
}

export interface ModelResponse {
  content: string;
  usage: {
    inputTokens: number;
    outputTokens: number;
    totalTokens: number;
  };
  model: string;
  finishReason: string;
}

export class LLMModelManager {
  private bedrockClient: BedrockRuntimeClient;
  private modelConfigs: Map<string, ModelConfig> = new Map();

  constructor() {
    this.bedrockClient = new BedrockRuntimeClient({
      region: process.env.AWS_BEDROCK_REGION || 'us-east-1'
    });
    
    this.initializeDefaultModels();
  }

  private initializeDefaultModels(): void {
    // Claude models
    this.modelConfigs.set('claude-3-sonnet', {
      provider: 'bedrock',
      modelId: 'anthropic.claude-3-sonnet-20240229-v1:0',
      maxTokens: 4096,
      temperature: 0.7
    });

    this.modelConfigs.set('claude-3-haiku', {
      provider: 'bedrock',
      modelId: 'anthropic.claude-3-haiku-20240307-v1:0',
      maxTokens: 4096,
      temperature: 0.7
    });

    // Nova models
    this.modelConfigs.set('nova-pro', {
      provider: 'bedrock',
      modelId: 'amazon.nova-pro-v1:0',
      maxTokens: 4096,
      temperature: 0.7
    });

    this.modelConfigs.set('nova-lite', {
      provider: 'bedrock',
      modelId: 'amazon.nova-lite-v1:0',
      maxTokens: 4096,
      temperature: 0.7
    });

    // OpenAI models (direct API)
    this.modelConfigs.set('gpt-4', {
      provider: 'openai',
      modelId: 'gpt-4',
      maxTokens: 4096,
      temperature: 0.7
    });

    this.modelConfigs.set('gpt-3.5-turbo', {
      provider: 'openai',
      modelId: 'gpt-3.5-turbo',
      maxTokens: 4096,
      temperature: 0.7
    });

    // Anthropic models (direct API)
    this.modelConfigs.set('claude-3-opus-direct', {
      provider: 'anthropic',
      modelId: 'claude-3-opus-20240229',
      maxTokens: 4096,
      temperature: 0.7
    });

    this.modelConfigs.set('claude-3-sonnet-direct', {
      provider: 'anthropic',
      modelId: 'claude-3-sonnet-20240229',
      maxTokens: 4096,
      temperature: 0.7
    });
  }

  async invoke(
    modelName: string,
    prompt: string,
    options?: Partial<ModelConfig>
  ): Promise<ModelResponse> {
    const config = this.modelConfigs.get(modelName);
    if (!config) {
      throw new Error(`Model not found: ${modelName}`);
    }

    const mergedConfig = { ...config, ...options };

    switch (config.provider) {
      case 'bedrock':
        return this.invokeBedrock(mergedConfig, prompt);
      case 'openai':
        return this.invokeOpenAI(mergedConfig, prompt);
      case 'anthropic':
        return this.invokeAnthropic(mergedConfig, prompt);
      default:
        throw new Error(`Unsupported provider: ${config.provider}`);
    }
  }

  private async invokeBedrock(
    config: ModelConfig,
    prompt: string
  ): Promise<ModelResponse> {
    const body = JSON.stringify({
      anthropic_version: 'bedrock-2023-05-31',
      max_tokens: config.maxTokens || 4096,
      temperature: config.temperature || 0.7,
      top_p: config.topP || 0.9,
      messages: [
        {
          role: 'user',
          content: prompt
        }
      ]
    });

    const command = new InvokeModelCommand({
      modelId: config.modelId,
      body,
      contentType: 'application/json',
      accept: 'application/json'
    });

    const response = await this.bedrockClient.send(command);
    const responseBody = JSON.parse(new TextDecoder().decode(response.body));

    return {
      content: responseBody.content[0].text,
      usage: {
        inputTokens: responseBody.usage.input_tokens,
        outputTokens: responseBody.usage.output_tokens,
        totalTokens: responseBody.usage.input_tokens + responseBody.usage.output_tokens
      },
      model: config.modelId,
      finishReason: responseBody.stop_reason
    };
  }

  private async invokeOpenAI(
    config: ModelConfig,
    prompt: string
  ): Promise<ModelResponse> {
    const { OpenAI } = await import('openai');
    const openai = new OpenAI({
      apiKey: process.env.OPENAI_API_KEY
    });

    const response = await openai.chat.completions.create({
      model: config.modelId,
      messages: [{ role: 'user', content: prompt }],
      max_tokens: config.maxTokens || 4096,
      temperature: config.temperature || 0.7,
      top_p: config.topP || 0.9
    });

    const choice = response.choices[0];
    return {
      content: choice.message.content || '',
      usage: {
        inputTokens: response.usage?.prompt_tokens || 0,
        outputTokens: response.usage?.completion_tokens || 0,
        totalTokens: response.usage?.total_tokens || 0
      },
      model: config.modelId,
      finishReason: choice.finish_reason || 'stop'
    };
  }

  private async invokeAnthropic(
    config: ModelConfig,
    prompt: string
  ): Promise<ModelResponse> {
    const { Anthropic } = await import('@anthropic-ai/sdk');
    const anthropic = new Anthropic({
      apiKey: process.env.ANTHROPIC_API_KEY
    });

    const response = await anthropic.messages.create({
      model: config.modelId,
      max_tokens: config.maxTokens || 4096,
      temperature: config.temperature || 0.7,
      messages: [{ role: 'user', content: prompt }]
    });

    const content = response.content[0];
    return {
      content: content.type === 'text' ? content.text : '',
      usage: {
        inputTokens: response.usage.input_tokens,
        outputTokens: response.usage.output_tokens,
        totalTokens: response.usage.input_tokens + response.usage.output_tokens
      },
      model: config.modelId,
      finishReason: response.stop_reason || 'end_turn'
    };
  }

  getAvailableModels(): string[] {
    return Array.from(this.modelConfigs.keys());
  }

  getModelConfig(modelName: string): ModelConfig | undefined {
    return this.modelConfigs.get(modelName);
  }

  registerModel(name: string, config: ModelConfig): void {
    this.modelConfigs.set(name, config);
  }
}

File: backend/src/llm/model_manager.py
from typing import Any, Dict, List, Optional

from .model_provider_abstract import ModelConfig, ModelProvider, ModelProviderFactory
from .providers.bedrock_provider import BedrockProvider
from .providers.openai_provider import OpenAIProvider


class ModelManager:
    """LLM 모델 관리자"""

    def __init__(self):
        self.providers: Dict[str, ModelProvider] = {}
        self.default_configs = self._load_default_configs()
        self._register_providers()

    def _register_providers(self):
        """프로바이더 등록"""
        ModelProviderFactory.register("openai", OpenAIProvider)
        ModelProviderFactory.register("bedrock", BedrockProvider)

    def _load_default_configs(self) -> Dict[str, ModelConfig]:
        """기본 모델 설정 로드"""
        return {
            "gpt-4": ModelConfig(
                name="gpt-4",
                provider="openai",
                max_tokens=4096,
                temperature=0.7,
                top_p=1.0,
                frequency_penalty=0.0,
                presence_penalty=0.0,
                stop_sequences=[],
            ),
            "claude-3-sonnet": ModelConfig(
                name="anthropic.claude-3-sonnet-20240229-v1:0",
                provider="bedrock",
                max_tokens=4096,
                temperature=0.7,
                top_p=1.0,
                frequency_penalty=0.0,
                presence_penalty=0.0,
                stop_sequences=[],
            ),
            "gpt-3.5-turbo": ModelConfig(
                name="gpt-3.5-turbo",
                provider="openai",
                max_tokens=4096,
                temperature=0.7,
                top_p=1.0,
                frequency_penalty=0.0,
                presence_penalty=0.0,
                stop_sequences=[],
            ),
        }

    async def get_provider(self, model_name: str) -> ModelProvider:
        """모델 프로바이더 가져오기"""
        if model_name not in self.providers:
            if model_name not in self.default_configs:
                raise ValueError(f"Unknown model: {model_name}")

            config = self.default_configs[model_name]
            provider = ModelProviderFactory.create(config.provider, config)
            await provider.initialize()
            self.providers[model_name] = provider

        return self.providers[model_name]

    async def generate(
        self, model_name: str, prompt: str, options: Optional[Dict[str, Any]] = None
    ):
        """텍스트 생성"""
        provider = await self.get_provider(model_name)
        return await provider.generate(prompt, options)

    async def stream_generate(
        self, model_name: str, prompt: str, options: Optional[Dict[str, Any]] = None
    ):
        """스트리밍 텍스트 생성"""
        provider = await self.get_provider(model_name)
        async for chunk in provider.stream_generate(prompt, options):
            yield chunk

    def list_models(self) -> List[str]:
        """사용 가능한 모델 목록"""
        return list(self.default_configs.keys())

    def get_model_info(self, model_name: str) -> Dict[str, Any]:
        """모델 정보 조회"""
        if model_name not in self.default_configs:
            raise ValueError(f"Unknown model: {model_name}")

        config = self.default_configs[model_name]
        return {
            "name": config.name,
            "provider": config.provider,
            "max_tokens": config.max_tokens,
            "temperature": config.temperature,
        }
